backend/src/apiserver/resource/resource_manager.go

```diff
@@ -683,6 +683,56 @@ func (r *ResourceManager) DeleteJob(jobID string) error {
 	return nil
 }
 
+func (r *ResourceManager) UpdateJob(apiJob *api.Job) (err error) {
+	defer func() {
+		err = util.Wrapf(err, "update job %q failed", apiJob.Name)
+	}()
+
+	job, err := r.checkJobExist(apiJob.Id)
+	if err != nil {
+		return err
+	}
+
+	cli := r.getScheduledWorkflowClient(job.Namespace)
+	swf, err := cli.Get(job.Name, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	w := util.Workflow{Workflow: &workflowapi.Workflow{
+		Spec: swf.Spec.Workflow.Spec,
+	}}
+
+	err = w.VerifyParameters(toParametersMap(apiJob.GetPipelineSpec().GetParameters()))
+	if err != nil {
+		return util.Wrap(err, "failed to verify new params")
+	}
+
+	swf.Spec.MaxConcurrency = &apiJob.MaxConcurrency
+	swf.Spec.NoCatchup = &apiJob.NoCatchup
+	swf.Spec.Trigger = *toCRDTrigger(apiJob.Trigger)
+	swf.Spec.Workflow.Parameters = toCRDParameter(apiJob.GetPipelineSpec().GetParameters())
+
+	if len(apiJob.GetServiceAccount()) > 0 {
+		swf.Spec.Workflow.Spec.ServiceAccountName = apiJob.GetServiceAccount()
+	}
+
+	job.DisplayName = apiJob.Name
+	job.Description = apiJob.Description
+
+	err = job.UpdateWorkflow(&util.ScheduledWorkflow{ScheduledWorkflow: swf})
```

Comment:
I am a little confused here. E.g., inside the job.UpdateWorkflow, the swf's maxConcurrency is used to set job's maxConcurrency. And several lines before job.UpdateWorkflow is called, the job's maxConcurrency is used to set swf's maxConcurrency. So we do this because?

Reply:
Yep so it all doesn't feel super clean, but we have 1. the apiJob coming from the user, 2. the job pulled from the db, and 3. the SWF from the workflow store. WE have to update both the swf and the job according to apiJob.

The apiJob applies updates to both the SWF and some properties of job which are not contained in the SWF. To make sure there are never any discrepancies between the SWF and the db job, it seemed right to 1. makes changes to the SWF, and 2. update the whole workflow of the job with that SWF, instead of applying changes to the two objects separately.

It generally feels quite complex though, this whole job/swf/apiJob dance. Open to better suggestions/changes to this design